# 分檔

- Name: **林啟允**
- Student ID: **C44056223**
- Final Edit Date: **2020/6/2**

## Outline

- [參考資料](#參考資料)
- [匯入檔案](#匯入檔案)
- [進行分檔](#進行分檔)
- [確認檔案是否分割成功](#確認檔案是否分割成功)

## 參考資料

- [I have a CSV file and I need to split into two CSV files. How can I achieve this through Python script?](https://www.quora.com/I-have-a-CSV-file-and-I-need-to-split-into-two-CSV-files-How-can-I-achieve-this-through-Python-script)

## 匯入檔案

- 先觀察資料筆數，再將其分為 9:1 的 Training Set 和 Testing Set

In [1]:
import pandas as pd

df = pd.read_csv('google_review_ratings.csv')

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5456 entries, 0 to 5455
Data columns (total 26 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   User         5456 non-null   object 
 1   Category 1   5456 non-null   float64
 2   Category 2   5456 non-null   float64
 3   Category 3   5456 non-null   float64
 4   Category 4   5456 non-null   float64
 5   Category 5   5456 non-null   float64
 6   Category 6   5456 non-null   float64
 7   Category 7   5456 non-null   float64
 8   Category 8   5456 non-null   float64
 9   Category 9   5456 non-null   float64
 10  Category 10  5456 non-null   float64
 11  Category 11  5456 non-null   object 
 12  Category 12  5455 non-null   float64
 13  Category 13  5456 non-null   float64
 14  Category 14  5456 non-null   float64
 15  Category 15  5456 non-null   float64
 16  Category 16  5456 non-null   float64
 17  Category 17  5456 non-null   float64
 18  Category 18  5456 non-null   float64
 19  Catego

## 進行分檔

- 由 `df.info()` 可以知道共有 5456 筆 entries，所以分為:
    - index 0 to 4909: `train.csv`，共 4910 筆
    - index 4910 to 5455: `test.csv`，共 546 筆，但會缺少資料欄位名稱，所以會額外再新增欄位名稱在第一列，總共 547 筆

In [3]:
import os
from os import path

# 確認檔案是否已存在
if path.exists('train.csv') == True:
    os.remove('train.csv')
if path.exists('test.csv') == True:
    os.remove('test.csv')

cond = 0
cut = 4910

# 將資料切成兩部分: train data : test data = 9 : 1
with open('google_review_ratings.csv', 'r', encoding='UTF-8') as file:
    for line in file:
        with open('train.csv', 'a', encoding='UTF-8') as file_1:
            with open('test.csv', 'a', encoding='UTF-8') as file_2:
                if cond < cut:
                    if cond == 0:
                        file_2.write(line + '\n')       # 在 test.csv 加入欄位名稱
                    file_1.write(line + '\n')
                else:
                    file_2.write(line + '\n')
        cond += 1

## 確認檔案是否分割成功

### train.csv

In [4]:
df_train = pd.read_csv('train.csv')
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4909 entries, 0 to 4908
Data columns (total 26 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   User         4909 non-null   object 
 1   Category 1   4909 non-null   float64
 2   Category 2   4909 non-null   float64
 3   Category 3   4909 non-null   float64
 4   Category 4   4909 non-null   float64
 5   Category 5   4909 non-null   float64
 6   Category 6   4909 non-null   float64
 7   Category 7   4909 non-null   float64
 8   Category 8   4909 non-null   float64
 9   Category 9   4909 non-null   float64
 10  Category 10  4909 non-null   float64
 11  Category 11  4909 non-null   object 
 12  Category 12  4908 non-null   float64
 13  Category 13  4909 non-null   float64
 14  Category 14  4909 non-null   float64
 15  Category 15  4909 non-null   float64
 16  Category 16  4909 non-null   float64
 17  Category 17  4909 non-null   float64
 18  Category 18  4909 non-null   float64
 19  Catego

In [5]:
df_train.head(10)

User  Category 1  Category 2  Category 3  Category 4  Category 5  \
0   User 1         0.0         0.0        3.63        3.65         5.0   
1   User 2         0.0         0.0        3.63        3.65         5.0   
2   User 3         0.0         0.0        3.63        3.63         5.0   
3   User 4         0.0         0.5        3.63        3.63         5.0   
4   User 5         0.0         0.0        3.63        3.63         5.0   
5   User 6         0.0         0.0        3.63        3.63         5.0   
6   User 7         0.0         5.0        3.63        3.63         5.0   
7   User 8         0.0         5.0        3.63        3.63         5.0   
8   User 9         0.0         5.0        3.64        3.64         5.0   
9  User 10         0.0         5.0        3.64        3.64         5.0   

   Category 6  Category 7  Category 8  Category 9  ...  Category 16  \
0        2.92        5.00        2.35        2.33  ...         0.59   
1        2.92        5.00        2.64        2.33  ...         0.59   
2        2.92        5.00        2.64        2.33  ...         0.59   
3        2.92        5.00        2.35        2.33  ...         0.59   
4        2.92        5.00        2.64        2.33  ...         0.59   
5        2.92        5.00        2.63        2.33  ...         0.59   
6        2.92        3.03        2.35        2.33  ...         0.59   
7        2.92        5.00        2.63        2.33  ...         0.60   
8        2.92        3.03        2.62        2.32  ...         0.60   
9        2.92        5.00        2.35        2.32  ...         0.59   

  Category 17  Category 18  Category 19  Category 20  Category 21  \
0         0.5          0.0          0.5          0.0          0.0   
1         0.5          0.0          0.5          0.0          0.0   
2         0.5          0.0          0.5          0.0          0.0   
3         0.5          0.0          0.5          0.0          0.0   
4         0.5          0.0          0.5          0.0          0.0   
5         0.5          0.0          0.5          0.0          0.0   
6         0.5          0.0          0.5          0.0          0.0   
7         0.5          0.0          0.5          0.0          0.0   
8         0.0          0.0          0.5          0.0          0.0   
9         0.0          0.0          0.0          0.0          0.0   

   Category 22  Category 23  Category 24  Unnamed: 25  
0          0.0          0.0          0.0          NaN  
1          0.0          0.0          0.0          NaN  
2          0.0          0.0          0.0          NaN  
3          0.0          0.0          0.0          NaN  
4          0.0          0.0          0.0          NaN  
5          0.0          0.0          0.0          NaN  
6          0.0          0.0          0.0          NaN  
7          0.0          0.0          0.0          NaN  
8          0.0          0.0          0.0          NaN  
9          0.0          0.0          0.0          NaN  

[10 rows x 26 columns]

### test.csv

In [6]:
df_test = pd.read_csv('test.csv')
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 547 entries, 0 to 546
Data columns (total 26 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   User         547 non-null    object 
 1   Category 1   547 non-null    float64
 2   Category 2   547 non-null    float64
 3   Category 3   547 non-null    float64
 4   Category 4   547 non-null    float64
 5   Category 5   547 non-null    float64
 6   Category 6   547 non-null    float64
 7   Category 7   547 non-null    float64
 8   Category 8   547 non-null    float64
 9   Category 9   547 non-null    float64
 10  Category 10  547 non-null    float64
 11  Category 11  547 non-null    float64
 12  Category 12  547 non-null    float64
 13  Category 13  547 non-null    float64
 14  Category 14  547 non-null    float64
 15  Category 15  547 non-null    float64
 16  Category 16  547 non-null    float64
 17  Category 17  547 non-null    float64
 18  Category 18  547 non-null    float64
 19  Category

In [7]:
df_test.head(10)

User  Category 1  Category 2  Category 3  Category 4  Category 5  \
0  User 4910        1.08        1.31        2.63        2.63        2.75   
1  User 4911        1.14        1.32        2.64        2.64        2.75   
2  User 4912        1.15        1.33        2.65        2.65        2.76   
3  User 4913        1.10        1.34        2.66        2.66        2.76   
4  User 4914        1.11        1.63        2.67        2.67        2.77   
5  User 4915        1.13        1.17        2.68        2.68        2.77   
6  User 4916        1.14        1.61        2.69        2.69        2.78   
7  User 4917        1.15        1.19        1.29        2.70        2.78   
8  User 4918        1.24        1.64        1.31        2.71        2.79   
9  User 4919        1.25        1.66        1.32        2.73        2.79   

   Category 6  Category 7  Category 8  Category 9  ...  Category 16  \
0        5.00         5.0        2.15        3.12  ...         0.69   
1        2.79         5.0        3.12        3.11  ...         0.69   
2        2.77         5.0        3.10        3.10  ...         0.69   
3        5.00         5.0        3.10        3.09  ...         0.68   
4        5.00         5.0        3.09        3.08  ...         0.68   
5        2.78         5.0        3.08        3.07  ...         0.68   
6        5.00         5.0        2.22        3.06  ...         0.67   
7        2.79         5.0        2.19        3.05  ...         0.67   
8        2.79         5.0        2.06        3.04  ...         0.67   
9        2.79         5.0        2.06        3.03  ...         0.66   

   Category 17  Category 18  Category 19  Category 20  Category 21  \
0         0.67         0.68         0.70         0.72         0.00   
1         0.67         0.67         0.69         0.00         0.00   
2         0.67         0.67         0.68         0.00         0.00   
3         0.66         0.66         0.68         3.16         0.00   
4         0.66         0.66         0.68         3.16         0.00   
5         0.65         0.65         0.58         0.70         0.00   
6         0.65         0.65         0.67         3.15         0.00   
7         0.64         0.65         0.66         0.56         0.51   
8         0.64         0.64         0.65         0.69         0.54   
9         0.64         0.63         0.65         0.68         0.54   

   Category 22  Category 23  Category 24  Unnamed: 25  
0         1.02         1.03         1.05          NaN  
1         1.03         1.04         1.06          NaN  
2         1.04         1.05         1.07          NaN  
3         1.05         1.06         1.08          NaN  
4         1.06         1.07         1.10          NaN  
5         1.08         1.08         1.18          NaN  
6         1.06         1.04         1.20          NaN  
7         1.02         1.05         1.21          NaN  
8         1.03         1.06         1.23          NaN  
9         0.58         1.07         1.24          NaN  

[10 rows x 26 columns]